File: lib/domain/core/decimal.ml
type t = int64

let scale = 8
let unit_ = 100_000_000L (* 10^8 *)

let zero = 0L
let one = unit_

let of_int n = Int64.mul (Int64.of_int n) unit_
let of_float f = Int64.of_float (f *. Int64.to_float unit_)
let to_float x = Int64.to_float x /. Int64.to_float unit_

let add = Int64.add
let sub = Int64.sub
let neg = Int64.neg
let abs = Int64.abs

let mul a b =
  (* (a * b) / unit, via Int128-style split to avoid overflow on typical prices. *)
  let open Int64 in
  let hi = div a unit_ and lo = rem a unit_ in
  add (mul hi b) (div (mul lo b) unit_)

let div a b =
  if Int64.equal b 0L then raise Division_by_zero
  else
    let open Int64 in
    let hi = div a b in
    let rem_ = sub a (mul hi b) in
    add (mul hi unit_) (div (mul rem_ unit_) b)

let compare = Int64.compare
let equal = Int64.equal
let min a b = if compare a b <= 0 then a else b
let max a b = if compare a b >= 0 then a else b

let is_positive x = compare x zero > 0
let is_negative x = compare x zero < 0
let is_zero x = equal x zero

let to_string x =
  let sign = if is_negative x then "-" else "" in
  let x = Int64.abs x in
  let whole = Int64.div x unit_ in
  let frac = Int64.rem x unit_ in
  if Int64.equal frac 0L then Printf.sprintf "%s%Ld" sign whole
  else
    let s = Printf.sprintf "%08Ld" frac in
    let len = String.length s in
    let trim =
      let i = ref (len - 1) in
      while !i >= 0 && s.[!i] = '0' do decr i done;
      String.sub s 0 (!i + 1)
    in
    Printf.sprintf "%s%Ld.%s" sign whole trim

let of_string s =
  let s = String.trim s in
  if s = "" then invalid_arg "Decimal.of_string: empty";
  let neg_, rest =
    if s.[0] = '-' then true, String.sub s 1 (String.length s - 1)
    else if s.[0] = '+' then false, String.sub s 1 (String.length s - 1)
    else false, s
  in
  let whole, frac =
    match String.index_opt rest '.' with
    | None -> rest, ""
    | Some i -> String.sub rest 0 i, String.sub rest (i + 1) (String.length rest - i - 1)
  in
  let frac =
    if String.length frac > scale then String.sub frac 0 scale
    else frac ^ String.make (scale - String.length frac) '0'
  in
  let w = if whole = "" then 0L else Int64.of_string whole in
  let f = if frac = "" then 0L else Int64.of_string frac in
  let v = Int64.add (Int64.mul w unit_) f in
  if neg_ then Int64.neg v else v

File: ui/src/app/decimal.ts
/**
 * Fixed-point decimal mirroring OCaml `Core.Decimal` (`strategy/lib/domain/core/decimal.ml`).
 *
 * Same scale (10^8), same parse rules, same canonical string output —
 * so a string produced by either side round-trips bit-exactly through
 * the other's parser. Backed by `BigInt`; the OCaml side uses `int64`.
 *
 * Use this at the HTTP boundary for monetary fields. Inside the UI
 * the chart libraries take `number`, so most consumers pass through
 * `toNumber()` after parsing.
 */
export class Decimal {
  static readonly SCALE = 8;
  static readonly UNIT = 100_000_000n;

  /** Construct via `Decimal.fromString` / `fromInt` / `fromNumber`. */
  private constructor(private readonly raw: bigint) {}

  static readonly ZERO = new Decimal(0n);
  static readonly ONE = new Decimal(Decimal.UNIT);

  static fromInt(n: number | bigint): Decimal {
    const b = typeof n === 'bigint' ? n : BigInt(n);
    return new Decimal(b * Decimal.UNIT);
  }

  /** Lossy: `n` is IEEE 754. Use only when the source is already a JS
   *  number (e.g. a `<input type="number">`); for anything that came
   *  off the wire use `fromString`. */
  static fromNumber(n: number): Decimal {
    if (!Number.isFinite(n)) {
      throw new Error(`Decimal.fromNumber: non-finite ${n}`);
    }
    return Decimal.fromString(n.toString());
  }

  /**
   * Parses `[-+]?\d+(\.\d+)?`. Truncates fractional part beyond
   * 8 digits (no rounding) and pads shorter ones with trailing zeros —
   * matches the OCaml implementation's behaviour.
   */
  static fromString(s: string): Decimal {
    const trimmed = s.trim();
    if (trimmed === '') {
      throw new Error('Decimal.fromString: empty');
    }
    let rest = trimmed;
    let neg = false;
    if (rest[0] === '-') {
      neg = true;
      rest = rest.slice(1);
    } else if (rest[0] === '+') {
      rest = rest.slice(1);
    }
    const dot = rest.indexOf('.');
    const wholeStr = dot < 0 ? rest : rest.slice(0, dot);
    let fracStr = dot < 0 ? '' : rest.slice(dot + 1);
    if (fracStr.length > Decimal.SCALE) {
      fracStr = fracStr.slice(0, Decimal.SCALE);
    } else {
      fracStr = fracStr + '0'.repeat(Decimal.SCALE - fracStr.length);
    }
    if (wholeStr === '' && fracStr === '0'.repeat(Decimal.SCALE)) {
      return Decimal.ZERO;
    }
    if (!/^\d*$/.test(wholeStr) || !/^\d*$/.test(fracStr)) {
      throw new Error(`Decimal.fromString: invalid ${JSON.stringify(s)}`);
    }
    const w = wholeStr === '' ? 0n : BigInt(wholeStr);
    const f = fracStr === '' ? 0n : BigInt(fracStr);
    const v = w * Decimal.UNIT + f;
    return new Decimal(neg ? -v : v);
  }

  /**
   * Canonical decimal string: no trailing zeros in the fractional
   * part, no decimal point if the value is integral, leading `-`
   * for negatives. Bit-exact with `Core.Decimal.to_string`.
   */
  toString(): string {
    const sign = this.raw < 0n ? '-' : '';
    const abs = this.raw < 0n ? -this.raw : this.raw;
    const whole = abs / Decimal.UNIT;
    const frac = abs % Decimal.UNIT;
    if (frac === 0n) return `${sign}${whole}`;
    const padded = frac.toString().padStart(Decimal.SCALE, '0');
    const trimmed = padded.replace(/0+$/, '');
    return `${sign}${whole}.${trimmed}`;
  }

  /** Lossy projection for chart libraries that take `number`. */
  toNumber(): number {
    return Number(this.toString());
  }

  /** Underlying scaled integer. Exposed for tests / debugging. */
  toRaw(): bigint {
    return this.raw;
  }

  add(b: Decimal): Decimal {
    return new Decimal(this.raw + b.raw);
  }
  sub(b: Decimal): Decimal {
    return new Decimal(this.raw - b.raw);
  }
  /** `(a * b) / unit`. BigInt has unbounded precision, so no need
   *  for the int64 hi/lo split that the OCaml side uses. */
  mul(b: Decimal): Decimal {
    return new Decimal((this.raw * b.raw) / Decimal.UNIT);
  }
  /** `(a * unit) / b`. Throws on division by zero. */
  div(b: Decimal): Decimal {
    if (b.raw === 0n) throw new RangeError('Decimal.div: division by zero');
    return new Decimal((this.raw * Decimal.UNIT) / b.raw);
  }
  neg(): Decimal {
    return new Decimal(-this.raw);
  }
  abs(): Decimal {
    return new Decimal(this.raw < 0n ? -this.raw : this.raw);
  }

  cmp(b: Decimal): -1 | 0 | 1 {
    return this.raw < b.raw ? -1 : this.raw > b.raw ? 1 : 0;
  }
  eq(b: Decimal): boolean {
    return this.raw === b.raw;
  }
  isZero(): boolean {
    return this.raw === 0n;
  }
  isPositive(): boolean {
    return this.raw > 0n;
  }
  isNegative(): boolean {
    return this.raw < 0n;
  }

  static min(a: Decimal, b: Decimal): Decimal {
    return a.cmp(b) <= 0 ? a : b;
  }
  static max(a: Decimal, b: Decimal): Decimal {
    return a.cmp(b) >= 0 ? a : b;
  }
}
